c_binding/lsm_local_disk.c

```diff
@@ -751,3 +502,238 @@ int lsm_local_disk_vpd83_get(const char *disk_path, char **vpd83,
+        if (lsm_err != NULL)
+            *lsm_err = LSM_ERROR_CREATE_PLUGIN_MSG(rc, err_msg);
+        if (link_type != NULL)
+            *link_type = LSM_LOCAL_DISK_LINK_TYPE_UNKNOWN;
+    }
+
+    return rc;
+}
+
+int lsm_local_disk_ident_led_set(const char *disk_path, lsm_error **lsm_err)
+{
+    return _ses_ctrl(disk_path, lsm_err, _SES_DEV_CTRL_RQST_IDENT,
+                     _SES_CTRL_SET);
+}
+
+int lsm_local_disk_ident_led_clear(const char *disk_path, lsm_error **lsm_err)
+{
+    return _ses_ctrl(disk_path, lsm_err, _SES_DEV_CTRL_RQST_IDENT,
+                     _SES_CTRL_CLEAR);
+}
+
+int lsm_local_disk_fault_led_set(const char *disk_path, lsm_error **lsm_err)
+{
+    return _ses_ctrl(disk_path, lsm_err, _SES_DEV_CTRL_RQST_FAULT,
+                     _SES_CTRL_SET);
+}
+
+int lsm_local_disk_fault_led_clear(const char *disk_path, lsm_error **lsm_err)
+{
+    return _ses_ctrl(disk_path, lsm_err, _SES_DEV_CTRL_RQST_FAULT,
+                     _SES_CTRL_CLEAR);
+}
+
+static int _ses_ctrl(const char *disk_path, lsm_error **lsm_err,
+                     int action, int action_type)
+{
+    int rc = LSM_ERR_OK;
+    char err_msg[_LSM_ERR_MSG_LEN];
+    char tp_sas_addr[_SG_T10_SPL_SAS_ADDR_LEN];
+    int fd = -1;
+
+    _lsm_err_msg_clear(err_msg);
+
+    _good(_check_null_ptr(err_msg, 2 /* arg_count */, disk_path, lsm_err),
+          rc, out);
+
+    _good(_sg_io_open_ro(err_msg, disk_path, &fd), rc, out);
+
+    /* TODO(Gris Ge): Add support of NVMe enclosure */
+
+    /* Try use sysfs first to get SAS address. */
+    if ((strlen(disk_path) > strlen("/dev/")) &&
+        (strncmp(disk_path, "/dev/", strlen("/dev/")) == 0) &&
+        (strncmp(disk_path + strlen("/dev/"), "sd", strlen("sd")) == 0))
+        _sysfs_sas_addr_get(disk_path + strlen("/dev/"), tp_sas_addr);
+
+    /* Find out the target port address via SCSI VPD device Identification
+     * page:
+     *  SPC-5 rev7 Table 487 — ASSOCIATION field
+     * TODO(Gris Ge): try _sysfs_vpd_pg83_data_get() first, if fail use SGIO.
+     */
+    if (tp_sas_addr[0] == '\0')
+        _good(_sg_tp_sas_addr_of_disk(err_msg, fd, tp_sas_addr), rc, out);
+
+    /* SEND DIAGNOSTIC
+     * SES-3, 6.1.3 Enclosure Control diagnostic page
+     * SES-3, Table 78 — Device Slot control element
+     */
+    _good(_ses_dev_slot_ctrl(err_msg, tp_sas_addr, action, action_type),
+          rc, out);
+
+ out:
+    if (rc != LSM_ERR_OK) {
+        if (lsm_err != NULL)
+            *lsm_err = LSM_ERROR_CREATE_PLUGIN_MSG(rc, err_msg);
+    }
+    return rc;
+}
+
+static int _sysfs_sas_addr_get(const char *blk_name, char *tp_sas_addr)
+{
+    int rc = -1;
+    char sysfs_sas_addr[_SYSFS_SAS_ADDR_LEN];
+    char *sysfs_sas_path = NULL;
+    ssize_t read_size = -1;
+
+    memset(sysfs_sas_addr, 0, _SYSFS_SAS_ADDR_LEN);
+    memset(tp_sas_addr, 0, _SG_T10_SPL_SAS_ADDR_LEN);
+
+    sysfs_sas_path = (char *)
+        malloc(sizeof(char) * (strlen("/sys/block//device/sas_address") +
```

Comment:
I might just be slow today, but is there a reason for the double slash between block and device?